core/arch/aarch64/codec.txt

```diff
@@ -945,17 +946,188 @@ x101101011000000000101xxxxxxxxxx  cls     wx0 : wx5
 
 # Data Processing - Scalar Floating-Point and Advanced SIMD
+# FABD
+0x101110110xxxxx000101xxxxxxxxxx     fabd dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011101x1xxxxx110101xxxxxxxxxx     fabd dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/sub/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/sub/fp/sisd
+
+# FABS (scalar)
+00011110xx100000110000xxxxxxxxxx     fabs float_reg0 : float_reg5
+
+# FACGE
+0x101110010xxxxx001011xxxxxxxxxx     facge dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011100x1xxxxx111011xxxxxxxxxx     facge dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp/sisd
+
+# FACGT
+0x101110110xxxxx001011xxxxxxxxxx     facgt dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011101x1xxxxx111011xxxxxxxxxx     facgt dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp/sisd
 
 # FADD (vector)
+0x001110010xxxxx000101xxxxxxxxxx     fadd dq0 : dq5 dq16 fsz16 # Armv8.2
 0x0011100x1xxxxx110101xxxxxxxxxx     fadd dq0 : dq5 dq16 fsz
-0x001110010xxxxx000101xxxxxxxxxx     fadd dq0 : dq5 dq16 fsz16
 
 # FADD (scalar)
-00011110xx1xxxxx001010xxxxxxxxxx fadd float_reg0 : float_reg5 float_reg16
+00011110xx1xxxxx001010xxxxxxxxxx     fadd float_reg0 : float_reg5 float_reg16
+
+# FADDP (vector)
+0x101110010xxxxx000101xxxxxxxxxx     faddp dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011100x1xxxxx110101xxxxxxxxxx     faddp dq0 : dq5 dq16 fsz
+
+# FCMEQ (register)
+0x001110010xxxxx001001xxxxxxxxxx     fcmeq dq0 : dq5 dq16 fsz16 # Armv8.2
+0x0011100x1xxxxx111001xxxxxxxxxx     fcmeq dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp/sisd
+
+# FCMGE (register)
+0x101110010xxxxx001001xxxxxxxxxx     fcmge dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011100x1xxxxx111001xxxxxxxxxx     fcmge dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp/sisd
+
+# FCMGT (register)
+0x101110110xxxxx001001xxxxxxxxxx     fcmgt dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011101x1xxxxx111001xxxxxxxxxx     fcmgt dq0 : dq5 dq16 fsz
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp16/sisd
+# Missing encoding: aarch64/instrs/vector/arithmetic/binary/uniform/cmp/fp/sisd
+
+# FDIV (vector)
+0x101110010xxxxx001111xxxxxxxxxx     fdiv dq0 : dq5 dq16 fsz16 # Armv8.2
+0x1011100x1xxxxx111111xxxxxxxxxx     fdiv dq0 : dq5 dq16 fsz
+
+# FDIV (scalar)
+00011110xx1xxxxx000110xxxxxxxxxx     fdiv float_reg0 : float_reg5 float_reg16
+
+# FMADD
+00011111xx0xxxxx0xxxxxxxxxxxxxxx     fmadd float_reg0 : float_reg5 float_reg16 float_reg10
```

Comment:
I spot-checked just FMADD and it seems the bit after the 5 1's (bit 23) can be 0 and not x?

Reply:
According to the A64 ISA XML [1], bits 22 and 23 of FMADD (and other FP instructions) encode the element type (00 = Single, 01 = double, 11 = Half). xx here means it is set by (at least) one of the operands, in that case that case float_reg0, float_reg5, float_reg16, float_reg10. The actual encoding of the type is done [here](https://github.com/DynamoRIO/dynamorio/pull/2951/files/873b400d991eaf5fa8b81c2f627ad0aa6d5121fb#diff-b30aeb7ea40f676cc26f5821bba3bbc9L1063). 

[1] https://developer.arm.com/products/architecture/a-profile/exploration-tools

Reply:
In the ARM Architecture Reference Manual for ARMv8, maybe old since from 2013 `ARM DDI 0487A.a (ID090413)`, the FMADD entry in C6.3.87 on page C6-924 says:

```
31 30 29 28 27 26 25 24 23 22 21 20    16 15 14    10 9    5 4    0
 0  0  0  1  1  1  1  1  0  x  0       Rm  0    Ra      Rn     Rd
                         type o1          o0
```

I.e., it is saying that the top bit of the type is always 0 -- meaning the half type is not supported. Yet encode_opnd_float_reg() will accept half.

Reply:
Ah yes. The half encodings were added in Arm v8.2. That leads us to an interesting question: How should we deal with that? I also realized that there is no comment indicating that the half version is 8.2.

Reply:
For x86 we do not treat processor revisions as first-class in the encoder/decoder: it will accept a newer encoding when running on an older processor (and deal with it for execution only when SIGILL is raised) so it doesn't seem so bad to keep it how you have it, with perhaps as you say a comment in encode_opnd_float_reg() saying that a half type on 8.1 is illegal.

Reply:
Great thanks for having a look. Will do that tomorrow.

Reply:
Xref the drcpusim tool [1] which has its own accounting of which encodings are valid on which processor because the DR ISA doesn't tell it.  Its use case is an app that contains different assembly sequences for different architectures (such as optimized multimedia code) and wants to test that it's using only legal instructions on some legacy processor.  (Would such a tool be useful for ARM as well or does ARM already provide such a tool?)

[1] http://dynamorio.org/docs/page_drcpusim.html